core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -565,16 +566,23 @@ ClientStream newSubstream(ClientStreamTracer.Factory tracerFactory, Metadata new
         builder.proxyDetector != null ? builder.proxyDetector : GrpcUtil.DEFAULT_PROXY_DETECTOR;
     this.retryEnabled = builder.retryEnabled && !builder.temporarilyDisableRetry;
     this.loadBalancerFactory = new AutoConfiguredLoadBalancerFactory(builder.defaultLbPolicy);
+    this.blockingExecutorHolder =
+        new ExecutorHolder(
+            checkNotNull(builder.blockingExecutorPool, "blockingExecutorPool"));
     this.nameResolverRegistry = builder.nameResolverRegistry;
-    this.nameResolverArgs = NameResolver.Args.newBuilder()
-        .setDefaultPort(builder.getDefaultPort())
-        .setProxyDetector(proxyDetector)
-        .setSynchronizationContext(syncContext)
-        .setServiceConfigParser(
-            new ScParser(
-                retryEnabled, builder.maxRetryAttempts, builder.maxHedgedAttempts,
-                loadBalancerFactory))
-        .build();
+    this.nameResolverArgs =
+        NameResolver.Args.newBuilder()
+            .setDefaultPort(builder.getDefaultPort())
+            .setProxyDetector(proxyDetector)
+            .setSynchronizationContext(syncContext)
+            .setServiceConfigParser(
+                new ScParser(
+                    retryEnabled,
+                    builder.maxRetryAttempts,
+                    builder.maxHedgedAttempts,
+                    loadBalancerFactory))
+            .setBlockingExecutor(builder.blockingExecutorPool.getObject())
```

Comment:
This getObject() is not returned; the holder is released, but blockingExecutorHolder.getExecutor() wasn't called so it won't return the object. That particular issue can be fixed by just not using ExecutorHolder for the blocking pool. But...

We probably _do_ want to delay allocating the executor until first use. We can make a `ExecutorHolderExecutor` (tehe) with:
```
public void execute(Runnable runnable) {
  holder.getExecutor().execute(runnable);
}
```

Things might be reworked to be less weird, but that's probably good enough for now.

Reply:
Whoops, yeah I need to get the instance from the ExecutorHolder.

But wouldn't the proposed solution imbalance the SharedResource refcounting? There would be a `SharedResourceHolder.get()` each time runnable is submitted to the executor, but `SharedResourceHolder.release()` only gets called once when the Channel is terminated.

So iiuc there actually needs to be some additional reference tracking the in the `ExecutorHolderExecutor` to ensure getExecutor() is only called for the first runnable.

Reply:
The ExecutorHolder doesn't need any balancing. It is just lazily grabbing a reference. So the first time getExecutor() is called it will get a reference, but the rest will use what it has.

Reply:
Gotcha, fixed.